compiler/src/dotty/tools/dotc/util/GenericHashSet.scala

```diff
@@ -0,0 +1,191 @@
+package dotty.tools.dotc.util
+
+import dotty.tools.uncheckedNN
+
+object GenericHashSet:
+
+  /** The number of elements up to which dense packing is used.
+   *  If the number of elements reaches `DenseLimit` a hash table is used instead
+   */
+  inline val DenseLimit = 8
+
+/** A hash set that allows some privileged protected access to its internals
+ *  @param  initialCapacity  Indicates the initial number of slots in the hash table.
+ *                           The actual number of slots is always a power of 2, so the
+ *                           initial size of the table will be the smallest power of two
+ *                           that is equal or greater than the given `initialCapacity`.
+ *                           Minimum value is 4.
+*  @param  capacityMultiple The minimum multiple of capacity relative to used elements.
+ *                           The hash table will be re-sized once the number of elements
+ *                           multiplied by capacityMultiple exceeds the current size of the hash table.
+ *                           However, a table of size up to DenseLimit will be re-sized only
+ *                           once the number of elements reaches the table's size.
+ */
+abstract class GenericHashSet[T](initialCapacity: Int = 8, capacityMultiple: Int = 2) extends MutableSet[T] {
+  import GenericHashSet.DenseLimit
+
+  protected var used: Int = _
+  protected var limit: Int = _
+  protected var table: Array[AnyRef | Null] = _
+
+  clear()
+
+  private def allocate(capacity: Int) =
+    table = new Array[AnyRef | Null](capacity)
+    limit = if capacity <= DenseLimit then capacity - 1 else capacity / capacityMultiple
+
+  private def roundToPower(n: Int) =
+    if n < 4 then 4
+    else if Integer.bitCount(n) == 1 then n
+    else 1 << (32 - Integer.numberOfLeadingZeros(n))
+
+  def clear(resetToInitial: Boolean): Unit =
+    used = 0
+    if resetToInitial then allocate(roundToPower(initialCapacity))
+    else java.util.Arrays.fill(table, null)
+
+  /** The number of elements in the set */
+  def size: Int = used
+
+  protected def isDense = limit < DenseLimit
+
+  /** Hashcode, by default a processed `x.hashCode`, can be overridden */
+  protected def hash(key: T): Int
+
+  /** Hashcode, by default `equals`, can be overridden */
+  protected def isEqual(x: T, y: T): Boolean
```

Comment:
The Scaladocs mention a default but the methods are abstract. That doesn't seem right.

Reply:
Good catch, I have now corrected these docs